wrapper/src/main/java/software/amazon/jdbc/hostlistprovider/monitoring/MonitoringRdsMultiAzHostListProvider.java

```diff
@@ -31,32 +30,40 @@ public class MonitoringRdsMultiAzHostListProvider extends MonitoringRdsHostListP
   public MonitoringRdsMultiAzHostListProvider(
       final Properties properties,
       final String originalUrl,
-      final HostListProviderService hostListProviderService,
+      final ServiceContainer serviceContainer,
       final String topologyQuery,
       final String nodeIdQuery,
       final String isReaderQuery,
-      final PluginService pluginService,
       final String fetchWriterNodeQuery,
       final String fetchWriterNodeColumnName) {
-    super(properties, originalUrl, hostListProviderService, topologyQuery, nodeIdQuery, isReaderQuery,
-        "", pluginService);
+    super(
+        properties,
+        originalUrl,
+        serviceContainer,
+        topologyQuery,
+        nodeIdQuery,
+        isReaderQuery,
+        "");
     this.fetchWriterNodeQuery = fetchWriterNodeQuery;
     this.fetchWriterNodeColumnName = fetchWriterNodeColumnName;
   }
 
   @Override
   protected ClusterTopologyMonitor initMonitor() {
-    return monitors.computeIfAbsent(this.clusterId,
-        (key) -> new MultiAzClusterTopologyMonitorImpl(
-            key, topologyCache, this.initialHostSpec, this.properties, this.pluginService,
-            this.hostListProviderService, this.clusterInstanceTemplate,
-            this.refreshRateNano, this.highRefreshRateNano, TOPOLOGY_CACHE_EXPIRATION_NANO,
+    return monitorService.runIfAbsent(MultiAzClusterTopologyMonitorImpl.class,
+        this.clusterId,
+        () -> new MultiAzClusterTopologyMonitorImpl(
+            this.serviceContainer,
```

Comment:
Hmm I'm not sure monitor should use serviceContainer that belongs to a particular wrapper connection. Monitor service should provide all necessary services (like ConnectionService) to a monitor instance.

Reply:
I was wondering about this as well. We can switch to passing the individual services like you are suggesting but that makes the already-long constructor list 4 parameters longer and I know the intent of the service container was to reduce the amount of constructor parameters around the driver. Both options have pros and cons but I'm fine with either option so let me know what you prefer.